### PR TITLE
config-service backend support for fetch all keys

### PR DESCRIPTION
**Pull request by dileepk7**
Added support to fetch all keys from bucket when "keys" in toml is empty or '/'
example toml file:
`keys: [  "/" ]
`

### CODE REVIEW

File: src/github.com/kelseyhightower/confd/backends/config-service/client.go [hunk 1]
```diff
@@ -57,7 +57,13 @@ func (c *Client) GetValues(keys []string) (map[string]string, error) {
 	for _, v := range keys {
 		bucketsKey := strings.Split(strings.TrimPrefix(v, "/"), "/")
 		buckets := strings.Split(bucketsKey[0], ",")
-		key := bucketsKey[1]
+		var key string
+		if len(bucketsKey) >= 2 {
```
Comment (anvinjain) [inline src/github.com/kelseyhightower/confd/backends/config-service/client.go]: Leaving key empty 
```suggestion
		if len(bucketsKey) >= 2 {
```

File: src/github.com/kelseyhightower/confd/backends/config-service/client.go [hunk 2]
```diff
@@ -57,7 +57,13 @@ func (c *Client) GetValues(keys []string) (map[string]string, error) {
...
+			key = bucketsKey[1]
+		} else {
+			//when key is not given, fetch all keys
+			key = ""
```
Comment (anvinjain) [inline src/github.com/kelseyhightower/confd/backends/config-service/client.go]: This will fetch all keys when one of the key is empty or just a "/" 
IMO, this can very well be a mistake on user part and also not very intuitive. Instead fetch all keys when toml has "*". Something like..
`keys = [ "*" ]`
`*` is not a valid key in config service so there should be no clash with existing key in a bucket

File: src/github.com/kelseyhightower/confd/backends/config-service/client.go [hunk 3]
```diff
@@ -66,28 +72,49 @@ func (c *Client) GetValues(keys []string) (map[string]string, error) {
...
-			valType := reflect.TypeOf(val).Kind()
-			if valType == reflect.Slice {
-				data, err := ffjson.Marshal(val)
-				if err != nil {
-				    log.Error("Failed decoding from JSON")
-				} else {
-					vars[key] = string(data[:])
+			var allKeysInBucket []string
+
+			if key == "" {
```
Comment (anvinjain) [inline src/github.com/kelseyhightower/confd/backends/config-service/client.go]: See above comment. check `key == "*"`

File: src/github.com/kelseyhightower/confd/backends/config-service/client.go [hunk 4]
```diff
@@ -66,28 +72,49 @@ func (c *Client) GetValues(keys []string) (map[string]string, error) {
...
-				continue;
-			}
-			valType := reflect.TypeOf(val).Kind()
-			if valType == reflect.Slice {
-				data, err := ffjson.Marshal(val)
-				if err != nil {
-				    log.Error("Failed decoding from JSON")
-				} else {
-					vars[key] = string(data[:])
+			var allKeysInBucket []string
```
Comment (anvinjain) [inline src/github.com/kelseyhightower/confd/backends/config-service/client.go]: rename to `requestedKeys` since it can either be all keys in the bucket or just a single key depending on what user has specified in toml

File: src/github.com/kelseyhightower/confd/backends/config-service/client.go [hunk 5]
```diff
@@ -66,28 +72,49 @@ func (c *Client) GetValues(keys []string) (map[string]string, error) {
...
+
+			if key == "" {
+				//when key is empty get all keys in a bucket,
+				bucketKeys := dynamicBucket.GetKeys()
+				allKeysInBucket = []string{}
+				for k, _ := range bucketKeys {
+					//Trim string prifx "/"
+					k = strings.TrimPrefix(k, "/")
```
Comment (anvinjain) [inline src/github.com/kelseyhightower/confd/backends/config-service/client.go]: why is this required?

File: src/github.com/kelseyhightower/confd/backends/config-service/client.go [hunk 6]
```diff
@@ -66,28 +72,49 @@ func (c *Client) GetValues(keys []string) (map[string]string, error) {
...
 			} else {
-				switch val.(type) {
-					case int,int64:
-					vars[key] = strconv.FormatInt(val.(int64), 64)
+				allKeysInBucket = []string{key}
+			}
+
+			//For each key in bucket store value in variable named 'vars'
+			for keyIdx := range allKeysInBucket {
+				keyName := allKeysInBucket[keyIdx]
```
Comment (anvinjain) [inline src/github.com/kelseyhightower/confd/backends/config-service/client.go]: can be replaced with `for _, keyName := range requestedKeys {`
also consider just using `k` instead of `keyName`

**Review by anvinjain — CHANGES_REQUESTED** [review]

File: src/github.com/kelseyhightower/confd/backends/config-service/client.go (hunk 5 above)
Comment (dileepk7) [reply src/github.com/kelseyhightower/confd/backends/config-service/client.go]: refer line no 58. previously there was a piece of code which is trimming "/" from keys, so i kept the same.

Comment (anvinjain) [reply src/github.com/kelseyhightower/confd/backends/config-service/client.go]: That is for a different reason. It trims the leading slash from prefix which is a bucketname

File: src/github.com/kelseyhightower/confd/backends/config-service/client.go (hunk 6 above)
Comment (dileepk7) [reply src/github.com/kelseyhightower/confd/backends/config-service/client.go]: updated

File: src/github.com/kelseyhightower/confd/backends/config-service/client.go (hunk 4 above)
Comment (dileepk7) [reply src/github.com/kelseyhightower/confd/backends/config-service/client.go]: updated

File: src/github.com/kelseyhightower/confd/backends/config-service/client.go [hunk 7]
```diff
@@ -57,7 +57,10 @@ func (c *Client) GetValues(keys []string) (map[string]string, error) {
 	for _, v := range keys {
 		bucketsKey := strings.Split(strings.TrimPrefix(v, "/"), "/")
 		buckets := strings.Split(bucketsKey[0], ",")
-		key := bucketsKey[1]
+		var key string
```
Comment (anvinjain) [inline src/github.com/kelseyhightower/confd/backends/config-service/client.go]: you don't need this if block. original code is sufficient

File: src/github.com/kelseyhightower/confd/backends/config-service/client.go [hunk 8]
```diff
@@ -66,28 +69,46 @@ func (c *Client) GetValues(keys []string) (map[string]string, error) {
...
-				    log.Error("Failed decoding from JSON")
-				} else {
-					vars[key] = string(data[:])
+			var requestedKeys []string
+
+			if key == "*" {
+				//when key is "*" get all keys in a bucket,
+				bucketKeys := dynamicBucket.GetKeys()
+				requestedKeys = []string{}
+				for k, _ := range bucketKeys {
```
Comment (anvinjain) [inline src/github.com/kelseyhightower/confd/backends/config-service/client.go:75]: can this loop be replaced by a simple append(requestedKeys, dynamicBucket.GetKeys()...)

**Review by anvinjain — CHANGES_REQUESTED** [review]

File: src/github.com/kelseyhightower/confd/backends/config-service/client.go (hunk 7 above)
Comment (dileepk7) [reply src/github.com/kelseyhightower/confd/backends/config-service/client.go]: original code was throwing "runtime error: index out of range" error when bucketsKey has only one element.

File: src/github.com/kelseyhightower/confd/backends/config-service/client.go (hunk 8 above)
Comment (dileepk7) [reply src/github.com/kelseyhightower/confd/backends/config-service/client.go:75]: dynamicBucket.GetKeys() returns array of interface{}, converting this to string in loop.

**Review by anvinjain — APPROVED** [review]
